File: src/main/java/com/vspk/bookrest/exception/handler/UserAuthenticationEntrypoint.java
package com.vspk.bookrest.exception.handler;

import com.fasterxml.jackson.databind.ObjectMapper;
import com.vspk.bookrest.payload.AuthApiError;
import org.springframework.security.core.AuthenticationException;
import org.springframework.security.web.AuthenticationEntryPoint;

import javax.servlet.ServletException;
import javax.servlet.http.HttpServletRequest;
import javax.servlet.http.HttpServletResponse;
import java.io.IOException;

public class UserAuthenticationEntrypoint implements AuthenticationEntryPoint {

    @Override
    public void commence(HttpServletRequest request, HttpServletResponse response, AuthenticationException ex)
            throws IOException, ServletException {

        response.setContentType("application/json");
        response.setStatus(HttpServletResponse.SC_UNAUTHORIZED);

        final var mapper = new ObjectMapper();
        var authApiError = new AuthApiError(ex.getMessage());

        mapper.writeValue(response.getOutputStream(), authApiError);
    }
}

File: src/main/java/com/vspk/bookrest/exception/handler/UserAuthenticationEntrypoint.kt
package com.vspk.bookrest.exception.handler

import com.fasterxml.jackson.databind.ObjectMapper
import com.vspk.bookrest.payload.AuthApiError
import java.io.IOException
import javax.servlet.ServletException
import javax.servlet.http.HttpServletRequest
import javax.servlet.http.HttpServletResponse
import org.springframework.security.core.AuthenticationException
import org.springframework.security.web.AuthenticationEntryPoint

class UserAuthenticationEntrypoint : AuthenticationEntryPoint {
    @Throws(IOException::class, ServletException::class)
    override fun commence(request: HttpServletRequest, response: HttpServletResponse, ex: AuthenticationException) {
        response.contentType = "application/json"
        response.status = HttpServletResponse.SC_UNAUTHORIZED
        val mapper = ObjectMapper()
        val authApiError = AuthApiError(ex.message)
        mapper.writeValue(response.outputStream, authApiError)
    }
}